drivers/sensor/kx022/kx022_trigger.c

```diff
@@ -0,0 +1,445 @@
+
+LOG_MODULE_DECLARE(KX022, CONFIG_SENSOR_LOG_LEVEL);
+
+static void kx022_gpio_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
+{
+	struct kx022_data *data = CONTAINER_OF(cb, struct kx022_data, gpio_cb);
+	
+#if defined(CONFIG_KX022_TRIGGER_OWN_THREAD)
+	k_sem_give(&data->trig_sem);
+#elif defined(CONFIG_KX022_TRIGGER_GLOBAL_THREAD)
+	k_work_submit(&data->work);
+#endif
+}
+
+static void kx022_handle_motion_int(const struct device *dev)
+{
+	struct kx022_data *data = dev->data;
+
+	if (data->motion_handler != NULL) {
+		data->motion_handler(dev, &data->motion_trigger);
+	}
+}
+
+static void kx022_handle_drdy_int(const struct device *dev)
+{
+	struct kx022_data *data = dev->data;
+
+	if (data->drdy_handler != NULL) {
+		data->drdy_handler(dev, &data->drdy_trigger);
+	}
+}
+static void kx022_handle_slope_int(const struct device *dev)
+{
+	struct kx022_data *data = dev->data;
+
+	if (data->slope_handler != NULL) {
+		data->slope_handler(dev, &data->slope_trigger);
+	}
+}
+
+static void kx022_handle_int(const struct device *dev)
+{
+	struct kx022_data *data = dev->data;
+	uint8_t status, clr;
+
+	if (data->hw_tf->read_reg(data, KX022_REG_INS2, &status) < 0) {
+		return;
+	}
+
+	if (status & KX022_MASK_INS2_WUFS) {
+		kx022_handle_motion_int(dev);
+	}
+
+	if (status & KX022_MASK_INS2_DRDY) {
+		kx022_handle_drdy_int(dev);
+		//k_msleep(50);
+	}
+
+	if (status & KX022_MASK_INS2_TPS) {
+		kx022_handle_slope_int(dev);
+	}
+
+	if(data->hw_tf->read_reg(data, KX022_REG_INT_REL, &clr) < 0) {
+		LOG_ERR("KX022 : Failed clear int report flag");
+	}
+}
+
+#ifdef CONFIG_KX022_TRIGGER_OWN_THREAD
+static void kx022_thread(struct kx022_data *data)
+{
+	while (1) {
+		k_sem_take(&data->trig_sem, K_FOREVER);
+		kx022_handle_int(data->dev);
+	}
+}
+#endif
+
+#ifdef CONFIG_KX022_TRIGGER_GLOBAL_THREAD
+static void kx022_work_cb(struct k_work *work)
+{
+	struct kx022_data *data = CONTAINER_OF(work, struct kx022_data, work);
+
+	kx022_handle_int(data->dev);
+}
+#endif
+
+/**************************************************************
+ * FUNCTION : kx022_trigger_init
+ * use to initialize trigger function
+ *
+ * ************************************************************/
+int kx022_trigger_init(const struct device *dev)
+{
+	uint8_t val;
+	struct kx022_data *data = dev->data;
+	const struct kx022_config *cfg = dev->config;
+
+	/* setup data ready gpio interrupt */
+	data->gpio = device_get_binding(cfg->irq_port);
+	if (data->gpio == NULL) {
+		LOG_ERR("KX022 : Cannot get pointer to %s device.", cfg->irq_port);
+		return -EINVAL;
+	}
+
+	if(gpio_pin_configure(data->gpio, cfg->irq_pin, GPIO_INPUT | cfg->irq_flags)) {
+		LOG_ERR("KX022 : Unable to configure GPIO pin %u", cfg->irq_pin);
+		return -EINVAL;
+	}
+
+	gpio_init_callback(&data->gpio_cb, kx022_gpio_callback, BIT(cfg->irq_pin));
+
+	if (gpio_add_callback(data->gpio, &data->gpio_cb) < 0) {
+		LOG_ERR("KX022 : Could not set gpio callback.");
+		return -EIO;
+	}
+
+	data->dev = dev;
```

Comment:
is this used anywhere?

Reply:
if miss this line, interrupt cannot trigger,
i check the accelerometer register value is all correct.
maybe  thread need it ?